Vault setup with Spire OIDC

sudo-bmitch opened the pull request:
This one changes the key storage from a secret to inside of vault. Vault is setup with the transit plugin to allow authorized users to sign data without ever giving out the private key. And for authentication, this adds Spire with OIDC. Spire authenticates the node is in the cluster, and the request is coming from the Tekton chains service account. That allows the chains operator to get a short lived SVID from Spire which it uses to authenticate to Vault using OIDC.

Spire is considering taking over the helm chart. See ([Spire #2652](https://github.com/spiffe/spire/issues/2652))

Signed-off-by: Brandon Mitchell <git@bmitch.net>

joshualucas84 commented:
@sudo-bmitch Brandon so when i run make setup-vault i get the following error.

![image](https://user-images.githubusercontent.com/5225184/148463717-8db88197-075f-47d9-bb7f-7e691e02f4bc.png)

sudo-bmitch commented:
> @sudo-bmitch Brandon so when i run make setup-vault i get the following error.
> 
> ![image](https://user-images.githubusercontent.com/5225184/148463717-8db88197-075f-47d9-bb7f-7e691e02f4bc.png)

Hopefully that's just a race condition. Check if the pods and services came up with `kubectl get all -n spire`.

joshualucas84 commented:
> > @sudo-bmitch Brandon so when i run make setup-vault i get the following error.
> > ![image](https://user-images.githubusercontent.com/5225184/148463717-8db88197-075f-47d9-bb7f-7e691e02f4bc.png)
> 
> Hopefully that's just a race condition. Check if the pods and services came up with `kubectl get all -n spire`.

It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?

sudo-bmitch commented:
> It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?

@joshualucas84 well there's a face palm moment. I'll get that added to the Makefile. Just need work some git surgery to shuffle around some other commits to get them out of this branch first.

sudo-bmitch commented:
> > It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?
> 
> @joshualucas84 well there's a face palm moment. I'll get that added to the Makefile. Just need work some git surgery to shuffle around some other commits to get them out of this branch first.

Added another entry to the Makefile and rebased on main. There's still some work left to make this happen. I can get some debug code over to you to try out spire + vault + cosign. What I'm looking at right now is whether Cosign or Chains should be updated to allow the Chains KMS integration to use the Spire access to generate a vault token on the fly, or if that's too specific and we need to add some kind of extensibility interface.

sudo-bmitch commented:
This one is still in work upstream. Sigstore changes are done, and chains work is developed, just finishing the testing code on that so we can get it merged.

bradbeck commented on platform/.gitignore:
```diff
@@ -0,0 +1,3 @@
+certs
+components/vault/root-token
+components/vault/unseal-key
```
```suggestion
components/vault/unseal-key

```

bradbeck commented on platform/components/tekton/chains/patch_spire.json:
```diff
@@ -0,0 +1,55 @@
+      }
+    }
+  }
+}
```
```suggestion
}

```

bradbeck commented on platform/vault/values.yaml:
```diff
@@ -0,0 +1,6 @@
+
+server:
+  affinity: ""
+  
```
```suggestion

```

bradbeck commented on platform/components/vault/values.yaml:
```diff
@@ -0,0 +1,31 @@
+
+  #     storage "file" {
+  #       path = "/vault/data"
+  #     }
```
```suggestion
  #     }

```

sudo-bmitch commented on platform/components/vault/values.yaml:
```diff
@@ -0,0 +1,29 @@
+      subPath: spire-ca.pem
+      readOnly: true
+
+  # standalone:
+  #   enabled: true
+  #   config: |
+  #     listener "tcp" {
+  #       address = "[::]:8200"
+  #       cluster_address = "[::]:8201"
+  #     }
+  #     storage "file" {
+  #       path = "/vault/data"
+  #     }
```
I was looking at that warning, but not sure of the best way to clear it. It should be indented under server, same level as affinity. I'll give the yaml linter a check to see if there's a better way I should be doing it.

bradbeck commented on platform/06-setup-certs.sh:
```diff
@@ -0,0 +1,57 @@
+  -key "$ca_key" \
+  -days 3650 \
+  -subj "/CN='Demo CA'" \
+  -addext "basicConstraints=CA:true" \
```
Looks like we need at least `LibreSSL 3.1.0`.

sudo-bmitch replied:
This is one of the areas that will be helped with cert manager.

bradbeck commented on platform/20-spire-setup.sh:
```diff
@@ -1,88 +1,46 @@
-  -f "${QUICKSTART_DIR}/server-configmap.yaml" \
-  -f "${QUICKSTART_DIR}/server-statefulset.yaml" \
-  -f "${QUICKSTART_DIR}/server-service.yaml"
+helm repo add sudo-bmitch https://sudo-bmitch.github.io/helm-charts
```
This line fails for me:
```bash
+ helm repo add sudo-bmitch https://sudo-bmitch.github.io/helm-charts
Error: looks like "https://sudo-bmitch.github.io/helm-charts" is not a valid chart repository or cannot be reached: Get "https://sudo-bmitch.github.io/helm-charts/index.yaml": dial tcp [2606:50c0:8002::153]:443: connect: no route to host
```

rgreinho commented:
Running `make quickstart` produces the following error for me:

```bash
Configuring helm...
"stable" already exists with the same configuration, skipping
"bitnami" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "bitnami" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
bash platform/06-setup-certs.sh
unknown option -addext
```

It looks like the `openssl req` subcommand on OSX (`macOS 12.3.1`) does not have the `-addext` flag. 

My openssl version is:

```bash
$ openssl version
LibreSSL 2.8.3
```

sudo-bmitch commented:
> ```
> bash platform/06-setup-certs.sh
> unknown option -addext
> ```

I think Brad ran into that too, and has a resolution in his comment: https://github.com/buildsec/frsca/pull/94#discussion_r887209446

This is a short term gotcha, I'm planning to replace that script with cert-manager.

rgreinho reviewed (dismissed):
I got it to deploy successfully and I was able to run the buildpacks pipeline without any issue 👌

bradbeck commented on platform/20-spire-setup.sh:
```diff
@@ -1,88 +1,46 @@
-  -selector k8s_sat:agent_ns:spire \
-  -selector k8s_sat:agent_sa:spire-agent \
+  -spiffeID spiffe://example.org/ns/spire/node/ssf \
+  -selector k8s_psat:cluster:ssf \
```
Seems like we should change `ssf` to `frsca` or something else.

bradbeck commented on platform/20-spire-setup.sh:
```diff
@@ -1,88 +1,46 @@
-  -selector k8s_sat:cluster:demo-cluster \
-  -selector k8s_sat:agent_ns:spire \
-  -selector k8s_sat:agent_sa:spire-agent \
+  -spiffeID spiffe://example.org/ns/spire/node/ssf \
```
Seems like we should change `ssf` to `frsca` or something else.

bradbeck commented on platform/26-vault-setup.sh:
```diff
@@ -0,0 +1,97 @@
+}
+path "transit/sign/ssf" {
+  capabilities = ["create", "read", "update"]
+}
+path "transit/sign/ssf/*" {
+  capabilities = ["read", "update"]
+}
+path "transit/verify/ssf" {
+  capabilities = ["create", "read", "update"]
+}
+path "transit/verify/ssf/*" {
+  capabilities = ["read", "update"]
+}
+EOF
+
+vault_exec read auth/jwt/role/spire-chains-controller >/dev/null 2>&1 || \
+vault_exec write auth/jwt/role/spire-chains-controller \
+  role_type=jwt \
+  user_claim=sub \
+  bound_audiences=TESTING \
+  bound_subject=spiffe://example.org/ns/tekton-chains/sa/tekton-chains-controller \
+  token_ttl=15m \
+  token_policies=spire-transit
+
+vault_exec read transit/keys/ssf >/dev/null 2>&1 || \
+vault_exec write transit/keys/ssf \
+  type=ecdsa-p521
+
+vault_exec read -format=json transit/keys/ssf \
+  | jq -r .data.keys.\"1\".public_key >"${GIT_ROOT}/platform/certs/ssf.pem"
+
+kubectl -n vault create configmap ssf-certs --from-file=ssf.pem="${GIT_ROOT}/platform/certs/ssf.pem" --dry-run=client -o=yaml | kubectl apply -f -
+
+COSIGN_KEY=$(kubectl get secret signing-secrets -n tekton-chains -o jsonpath='{.data.cosign\.key}' 2> /dev/null || true)
+if [ -n "${COSIGN_KEY}" ]; then
+  # remove signing-secrets if created by cosign
+  kubectl -n tekton-chains delete secret signing-secrets
+fi
+( kubectl -n tekton-chains create secret generic signing-secrets \
+    --from-file=cosign.pub="${GIT_ROOT}/platform/certs/ssf.pem" \
```
Seems like we should change `ssf` to `frsca` or something else.

bradbeck commented on platform/components/spire/values.yaml:
```diff
@@ -0,0 +1,17 @@
+server:
+  trustDomain: "example.org"
+  clusterName: ssf
```
Seems like we should change `ssf` to `frsca` or something else.

bradbeck commented on platform/components/tekton/chains/patch_config_kms.yaml:
```diff
@@ -0,0 +1,9 @@
+data:
+  artifacts.oci.signer: kms
+  artifacts.taskrun.signer: kms
+  signers.kms.kmsref: "hashivault://ssf"
```
Seems like we should change `ssf` to `frsca` or something else.

sudo-bmitch commented:
@bradbeck ssf references have been updated to frsca.
